joomla/joomla-cms
# [4.0] Added additional TinyMCE 5 configuration option for "external_plugins"

### Summary of Changes
Added support for Tiny's "external_plugins" configuration option. This has been done using a repeatable pairing of plugin names and paths. 

External Plugins allow developers to use their own custom plugins, that live outside of Tiny's source folders, to help improve the content author's experience through adding of new tasks or custom workflows. They really open up the potential for developers to more easily integrate their own plugins with Joomla 4, especially given the power of what TinyMCE 5 can do.

Using this option is 100% optional. There is no loss of functionality if this configuration option is not set by the administrator.


### Testing Instructions
Build and install Joomla from the source as per usual.

When editing the TinyMCE Editor Plugin, a new field is available at the bottom of the main "Plugin" screen. This allows external plugins to be added at a per-set basis.

If used, the plugin's button can be added (albeit crudely) to the toolbar by using the Custom Button option.

You can get the source for a trivial plugin at:
https://github.com/martyf/tiny-plugin-helloworld

Steps to implement are:

1. Place the plugin JS file somewhere (could vary based on how the dev wants to set it all up - for example, "/tiny-plugins/helloworld.js"
2. Go to the Joomla 4 Admin, and TinyMCE Plugin settings, and select the Set (such as 0 for Admin). Enter a new External Plugin (bottom):
- **Name:** helloworld
- **Path:** /tiny-plugins/helloworld.js
3. In "custom buttons", enter the plugin button - in this case, `helloworld`.
4. Create an Article, and you should see a "Hello Button" in the Tiny toolbar
5. Click the button to trigger the plugin - this is a totally trivial example that demonstrates what Tiny 5 can do with its UI Components library (new with v5) - select an option, and on selecting it, will insert the markup. Totally trivial, but is an example plugin.

It is also worth checking that when you have no external plugins configured, the editing process still works. You can:
1. Remove the External Plugin configuration from the TinyMCE Plugin settings, and remove from "Custom Buttons", then Save and Close.
2. Create/Edit an Article, and the editor loads without error or warning.


### Documentation Changes Required

[TinyMCE editor plugin documentation](https://help.joomla.org/proxy?keyref=Help40:Extensions_Plugin_Manager_Edit&lang=en#Editor_-_TinyMCE "https://help.joomla.org/proxy?keyref=Help40:Extensions_Plugin_Manager_Edit&lang=en#Editor_-_TinyMCE") needs to be updated.

(added by @richard67)
